compiler/rustc_mir_build/src/builder/scope.rs

```diff
@@ -941,6 +972,73 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         self.cfg.start_new_block().unit()
     }
 
+    fn static_pattern_match(
+        &self,
+        cx: &RustcPatCtxt<'_, 'tcx>,
+        value: ExprId,
+        arms: &[ArmId],
+        built_match_tree: &BuiltMatchTree<'tcx>,
+    ) -> Option<BasicBlock> {
+        let it = arms.iter().zip(built_match_tree.branches.iter());
+        for (&arm_id, branch) in it {
+            let pat = cx.lower_pat(&*self.thir.arms[arm_id].pattern);
+
+            if let rustc_pattern_analysis::rustc::Constructor::Or = pat.ctor() {
+                for pat in pat.iter_fields() {
+                    let sub_branch = branch
+                        .sub_branches
+                        .get(pat.idx)
+                        .or_else(|| branch.sub_branches.last())
+                        .unwrap();
+
+                    match self.static_pattern_match_help(value, &pat.pat) {
+                        true => return Some(sub_branch.success_block),
+                        false => continue,
+                    }
+                }
+            } else if self.static_pattern_match_help(value, &pat) {
+                return Some(branch.sub_branches[0].success_block);
+            }
+        }
+
+        None
+    }
+
+    fn static_pattern_match_help(&self, value: ExprId, pat: &DeconstructedPat<'_, 'tcx>) -> bool {
+        use rustc_pattern_analysis::rustc::Constructor;
+
+        match pat.ctor() {
+            Constructor::Variant(variant_index) => match &self.thir[value].kind {
+                rustc_middle::thir::ExprKind::Adt(value_adt) => {
+                    return *variant_index == value_adt.variant_index;
+                }
+                other => todo!("{other:?}"),
+            },
+            Constructor::IntRange(int_range) => match &self.thir[value].kind {
+                rustc_middle::thir::ExprKind::Literal { lit, neg } => match &lit.node {
+                    LitKind::Int(n, _) => {
+                        let n = if *neg {
+                            let bits = pat.ty().primitive_size(self.tcx).bits();
+                            MaybeInfiniteInt::new_finite_int(
+                                (n.get() as i128).overflowing_neg().0 as u128,
+                                bits,
+                            )
+                        } else {
+                            MaybeInfiniteInt::new_finite_uint(n.get())
+                        };
+
+                        return IntRange::from_singleton(n).is_subrange(int_range);
+                    }
+
+                    other => todo!("{other:?}"),
+                },
+                other => todo!("{other:?}"),
+            },
+            Constructor::Wildcard => return true,
+            _ => false,
+        }
+    }
```

Comment:
`Maybe `static_pattern_match` and `static_pattern_match_help` should be methods on `BuiltMatchTree` to avoid having to expose the entire guts of `BuiltMatchTree` outside of the match lowering implementation?

Reply:
they are now methods on `Builder` in the file that defines `BuiltMatchTree`, so less `pub` is needed